src/main/org/codehaus/groovy/classgen/asm/OperandStack.java

```diff
@@ -336,14 +336,12 @@ private void doConvertAndCast(ClassNode targetType, boolean coerce) {
                 return;
             }
             box();
-        } else if (primTop) {
-            // top is primitive, target is not
-            // so box and do groovy cast
-            controller.getInvocationWriter().castToNonPrimitiveIfNecessary(top, targetType);
         } else if (primTarget) {
             // top is not primitive so unbox
             // leave that BH#doCast later
         } else {
+            // top is primitive, target is not
+            // so box and do groovy cast
```

Comment:
I am thinking this comment should be improved. It covers both the case the comment mentions and the default case. Let's see if Jochen or anyone else has thoughts on what that default branch is supposed to do with respect to how best to rephrase if needed.